inductiva/resources/machines_base.py

```diff
@@ -0,0 +1,214 @@
+"""Base class for machine groups."""
+import time
+
+from absl import logging
+
+import inductiva
+import inductiva.client.models
+from inductiva import api
+from inductiva.client.apis.tags import instance_api
+
+
+class BaseMachineGroup():
+    """Base class to manage Google Cloud resources."""
+    def __init__(
+        self,
+        machine_type: str,
+        spot: bool = False,
+        disk_size_gb: int = 40,
+        zone: str = "europe-west1-b"
+    ) -> None:
+        """Create a MachineGroup object.
+
+        Args:
+            machine_type: The type of GC machine to launch. Ex: "e2-standard-4".
+            Check https://cloud.google.com/compute/docs/machine-resource for
+            more information about machine types.
+            num_machines: The number of virtual machines to launch.
+            spot: Whether to use spot machines.
+            disk_size_gb: The size of the disk in GB, recommended min. is 40 GB.
+            zone: The zone where the machines will be launched.
+        """
+        self.machine_type = machine_type
+        self.spot = spot
+        self.disk_size_gb = disk_size_gb
+        self.zone = zone
+        self._id = None
+        self._name = None
+        self.create_time = None
+        self._started = False
+        self.register = True
+        # Set the API configuration that carries the information from the client
+        # to the backend.
+        self._api = instance_api.InstanceApi(api.get_client())
+        self._estimated_cost = None
+
+    @property
+    def id(self):
+        return self._id
+
+    @property
+    def name(self):
+        return self._name
+
+    def _register_machine_group(self, **kwargs):
+        instance_group_config = inductiva.client.models.InstanceGroupCreate(
+            machine_type=self.machine_type,
+            spot=self.spot,
+            disk_size_gb=self.disk_size_gb,
+            zone=self.zone,
+            **kwargs,
+        )
+        logging.info("Registering machine group configurations:")
+        resp = self._api.register_instance_group(body=instance_group_config)
+        self._id = resp.body["id"]
+        self._name = resp.body["name"]
+        self._log_machine_group_info()
+
+    @classmethod
+    def from_api_response(cls, resp: dict):
+        """Creates a MachineGroup object from an API response."""
+
+        machine_group = cls(
+            machine_type=resp["machine_type"],
+            spot=bool(resp["spot"]),
+            disk_size_gb=resp["disk_size_gb"],
+            zone=resp["zone"],
+        )
+        machine_group._id = resp["id"]
+        machine_group._name = resp["name"]
+        machine_group.create_time = resp["create_time"]
+        machine_group._started = True
```

Comment:
It doesn't make much sense to do it from the backend, because the backend wouldn't return a deleted or non-started machines.

Reply:
Why? Don't you need to verify if they are actually started, based on a single source of truth?

Reply:
This method is called, and designed to be called, only in one case to retrieve the outputs of the request to the api, and that request since it queries google for the active machines it only returns the ones that are already started.

Reply:
My only concern is that this is that a user can change it on his side and be able to launch again. However, we have methods to prevent the launching of to many machines, so maybe I am overcomplicating.